source/postprocess/heating_statistics.cc

```diff
@@ -0,0 +1,177 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#include <aspect/postprocess/heating_statistics.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    template <int dim>
+    std::pair<std::string,std::string>
+    HeatingStatistics<dim>::execute (TableHandler &statistics)
+    {
+      // create a quadrature formula based on the temperature element alone.
+      const QGauss<dim> quadrature_formula (this->get_fe().base_element(this->introspection().base_elements.temperature).degree+1);
+      const unsigned int n_q_points = quadrature_formula.size();
+
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values   |
+                               update_gradients |
+                               update_quadrature_points |
+                               update_JxW_values);
+
+      typename MaterialModel::Interface<dim>::MaterialModelInputs in(fe_values.n_quadrature_points, this->n_compositional_fields());
+      typename MaterialModel::Interface<dim>::MaterialModelOutputs out(fe_values.n_quadrature_points, this->n_compositional_fields());
+
+      in.strain_rate.resize(0); // we do not need the viscosity
+      std::vector<std::vector<double> > composition_values (this->n_compositional_fields(),std::vector<double> (quadrature_formula.size()));
+
+      std::list<std_cxx11::shared_ptr<HeatingModel::Interface<dim> > > heating_model_objects = this->get_heating_model_manager().get_heating_models();
+      const std::vector<std::string> heating_model_names = this->get_heating_model_manager().get_active_heating_model_names();
+
+      HeatingModel::HeatingModelOutputs heating_model_outputs(n_q_points, this->n_compositional_fields());
+
+      std::ostringstream output;
+      output.precision(4);
+
+      double average_heating_integral = 0.0;
+      double total_heating_integral = 0.0;
+
+      unsigned int index = 0;
+      for (typename std::list<std_cxx11::shared_ptr<HeatingModel::Interface<dim> > >::const_iterator
+           heating_model = heating_model_objects.begin();
+           heating_model != heating_model_objects.end(); ++heating_model, ++index)
```

Comment:
You have the order of loops wrong here. The expensive part is `fe_values.reinit()` which you currently have in a loop
```
  for (heating_model = ...)
    for (cell=...)
      {
         fe_values.reinit (cell);
         ...
```
You can make all of this much cheaper if you do

```
    for (cell=...)
      {
         fe_values.reinit (cell);
         for (heating_model = ...)
           ...
```
If you do it right, you also only have to have one global communication instead of one per heating model.